pp.c

```diff
@@ -7934,64 +7934,168 @@ PP(pp_multiparam)
+        /* Initially mark the pad slot for every named parameter variable as
+         * PADSTALE, so afterwards we can detect missing ones
+         */
+        for(size_t namedix = 0; namedix < n_named; namedix++) {
+            struct op_multiparam_named_aux *named = aux->named + namedix;
+            SvPADSTALE_on(PAD_SVl(named->padix));
         }
 
-        assert((argc % 2) == 0);
-
         while(argc) {
             SV **svp;
 
             svp = av_fetch(defav, parami, FALSE); parami++;
-            SV *key = svp ? *svp : &PL_sv_undef;
+            SV *name = svp ? *svp : &PL_sv_undef;
             svp = av_fetch(defav, parami, FALSE); parami++;
             SV *val = svp ? *svp : &PL_sv_undef;
             argc -= 2;
 
-            if (UNLIKELY(SvGMAGICAL(key)))
-                key = sv_mortalcopy(key);
+            STRLEN namelen;
+            const char *namepv = SvPV(name, namelen);
+
+            U32 namehash;
+            PERL_HASH(namehash, namepv, namelen);
+
+            PADOFFSET padix = 0;
+            for(size_t namedix = 0; namedix < n_named; namedix++) {
+                struct op_multiparam_named_aux *named = aux->named + namedix;
+
+                if(named->namehash != namehash || named->namelen != namelen)
+                    continue;
+                if(!strnEQ(named->namepv, namepv, namelen))
+                    continue;
+
+                padix = named->padix;
+                break;
+            }
+
+            if(padix) {
+                SV **padentry = &PAD_SVl(padix);
+                save_clearsv(padentry);
+
+                assert(TAINTING_get || !TAINT_get);
+                if (UNLIKELY(TAINT_get) && !SvTAINTED(val))
+                    TAINT_NOT;
+
+                SvPADSTALE_off(*padentry);
+                SvSetMagicSV(*padentry, val);
+            }
+            else if(slurpypad && slurpy == '@') {
+                AV *av = (AV *)*slurpypad;
 
-            hv_store_ent(hv, key, newSVsv(val), 0);
-            if (UNLIKELY(TAINT_get) && !SvTAINTED(val))
-                TAINT_NOT;
+                assert(TAINTING_get || !TAINT_get);
+                if (UNLIKELY(TAINT_get) && !SvTAINTED(name))
+                    TAINT_NOT;
+                if (UNLIKELY(TAINT_get) && !SvTAINTED(val))
+                    TAINT_NOT;
+
+                av_push(av, newSVsv(name));
+                av_push(av, newSVsv(val));
+            }
+            else if(slurpypad && slurpy == '%') {
+                HV *hv = (HV *)*slurpypad;
+
+                if (UNLIKELY(SvGMAGICAL(name)))
+                    name = sv_mortalcopy(name);
+
+                if (UNLIKELY(TAINT_get) && !SvTAINTED(val))
+                    TAINT_NOT;
+                hv_store_ent(hv, name, newSVsv(val), 0);
+            }
+            else {
+                // TODO: Consider collecting up all the names of unrecognised
+                // in one string
+                croak("Unrecognized named parameter '%s' to subroutine '%" SVf "'",
```

Comment:
If you change these croaks to Perl_croak_caller() the errors will include the line number of the caller rather than where the subroutine is defined which would be way more helpful (and in line with what signatures do without named parameters)

Reply:
`namepv` may or may not be utf8 encoded, and this message doesn't account for that.
```
croak_caller("Unrecognized named parameter '" UTF8f "' to subroutine '%" SVf "'",
                    UTF8fARG(SvUTF8(name), namepv), S_find_runcv_name());
```
If the above changes to fetch the UTF8 name this should pass `true` instead of `SvUTF8(name)`.